### PR TITLE
Update dotnet download Installer Url and Added a fallback check

### PR DESCRIPTION
**Pull request by sanjuyadav24**
**Task name**: DotNetCoreInstaller, UseDotNet

**Description**: Updated dotnet download urls

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N
**Attached related issue:** (Y/N) N

**Checklist**:
- [X] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [X] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.ps1 [hunk 1]
```diff
@@ -484,12 +487,21 @@ try {
...
+    try {
+        Say "Cannot download: $DownloadLink"
+        $DownloadLink = $FallbackDownloadLink
+        $ZipPath = [System.IO.Path]::GetTempFileName()
+        Say-Verbose "Fallabck zip path: $ZipPath"
+        Say "Downloading fallback link: $FallbackDownloadLink"
+        DownloadFile -Uri $FallbackDownloadLink -OutPath $ZipPath
+    } catch {
+        Say "Cannot download: $DownloadLink"
+        $DownloadLink = $LegacyDownloadLink
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.ps1:499]: minor should be '$LegacyDownloadLink'

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.ps1 [hunk 2]
```diff
@@ -484,12 +487,21 @@ try {
...
-    DownloadFile -Uri $DownloadLink -OutPath $ZipPath
+    try {
+        Say "Cannot download: $DownloadLink"
+        $DownloadLink = $FallbackDownloadLink
+        $ZipPath = [System.IO.Path]::GetTempFileName()
+        Say-Verbose "Fallabck zip path: $ZipPath"
+        Say "Downloading fallback link: $FallbackDownloadLink"
+        DownloadFile -Uri $FallbackDownloadLink -OutPath $ZipPath
+    } catch {
+        Say "Cannot download: $DownloadLink"
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.ps1:498]: minor : should be "Cannot download: $FallbackDownloadLink"

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh [hunk 3]
```diff
@@ -654,7 +658,19 @@ install_dotnet() {
     say_verbose "Zip path: $zip_path"
 
     say "Downloading link: $download_link"
-    download "$download_link" $zip_path || download_failed=true
+    download "$download_link" $zip_path || fallback_download_failed=true
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:661]: ? This should be "download_failed=true"

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh [hunk 4]
```diff
@@ -654,7 +658,19 @@ install_dotnet() {
...
+
+    say_verbose "download_failed: $download_failed"
+    say_verbose "valid_legacy_download_link: $valid_legacy_download_link"
+
+    if [ "$fallback_download_failed" = true ]; then
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:666]: This should be $download_failed ?

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh [hunk 5]
```diff
@@ -654,7 +658,19 @@ install_dotnet() {
...
+        say "Cannot download: $download_link"
+        download_link=$fallback_download_link
+        zip_path=$(mktemp $temporary_file_template)
+        say_verbose "Fallback zip path: $zip_path"
+        say "Downloading fallback link: $download_link"
+        download "$download_link" $zip_path || download_failed=true
+    fi
 
     #  if the download fails, download the legacy_download_link
     if [ "$download_failed" = true ] && [ "$valid_legacy_download_link" = true ]; then
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:676]: This should be $fallback_download_failed  ?

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh [hunk 6]
```diff
@@ -654,7 +658,19 @@ install_dotnet() {
...
+    say_verbose "download_failed: $download_failed"
+    say_verbose "valid_legacy_download_link: $valid_legacy_download_link"
+
+    if [ "$fallback_download_failed" = true ]; then
+        say "Cannot download: $download_link"
+        download_link=$fallback_download_link
+        zip_path=$(mktemp $temporary_file_template)
+        say_verbose "Fallback zip path: $zip_path"
+        say "Downloading fallback link: $download_link"
+        download "$download_link" $zip_path || download_failed=true
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:672]: This should be fallback_download_failed=true?

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 3 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:661]: there is a legacy download link, so I added two variables 
I need to swap the names fallback_download_failed and download_failed

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh [hunk 7]
```diff
@@ -654,7 +658,19 @@ install_dotnet() {
     say_verbose "Zip path: $zip_path"
 
     say "Downloading link: $download_link"
-    download "$download_link" $zip_path || download_failed=true
+    download "$download_link" $zip_path || fallback_download_failed=true
+
+    say_verbose "download_failed: $download_failed"
+    say_verbose "valid_legacy_download_link: $valid_legacy_download_link"
```
Comment (merlynomsft) [inline Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh:664]: I think this should move down to above "#  if the download fails, download the legacy_download_link"

File: Tasks/UseDotNetV2/versioninstaller.ts [hunk 8]
```diff
@@ -38,10 +38,14 @@ export class VersionInstaller {
                 var downloadPath = await toolLib.downloadToolWithRetries(downloadUrl)
             }
             catch (ex) {
-                tl.warning(tl.loc("CouldNotDownload", downloadUrl, ex));
-                let fallBackUrl = `https://dotnetcli.azureedge.net/dotnet/${this.packageType === "runtime" ? "Runtime" : "Sdk"}/${version}/${downloadUrl.substring(downloadUrl.lastIndexOf('/') + 1)}`;
-                console.log("Using fallback url for download: " + fallBackUrl);
-                var downloadPath = await toolLib.downloadToolWithRetries(fallBackUrl)
+                try {
+                    console.log("download exception");
```
[the remark this answers is not in the document]
Comment (sanjuyadav24) [reply Tasks/UseDotNetV2/versioninstaller.ts]: removed

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 3 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh]: updated the variables name

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 4 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh]: updated

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 5 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh]: updated

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 6 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh]: updated

File: Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh (hunk 7 above)
Comment (sanjuyadav24) [reply Tasks/DotNetCoreInstallerV0/externals/install-dotnet.sh]: updated

File: Tasks/UseDotNetV2/versioninstaller.ts [hunk 9]
```diff
@@ -38,10 +38,13 @@ export class VersionInstaller {
...
             catch (ex) {
-                tl.warning(tl.loc("CouldNotDownload", downloadUrl, ex));
-                let fallBackUrl = `https://dotnetcli.azureedge.net/dotnet/${this.packageType === "runtime" ? "Runtime" : "Sdk"}/${version}/${downloadUrl.substring(downloadUrl.lastIndexOf('/') + 1)}`;
-                console.log("Using fallback url for download: " + fallBackUrl);
-                var downloadPath = await toolLib.downloadToolWithRetries(fallBackUrl)
+                try {
+                    tl.warning(tl.loc("CouldNotDownload", downloadUrl, ex));
+                    var downloadPath = await this.downloadFromFallbackUrl("https://builds.dotnet.microsoft.com/dotnet", this.packageType, version, downloadUrl);
+                } catch(ex) {
+                    tl.warning(tl.loc("CouldNotDownload", downloadUrl, ex));
```
Comment (merlynomsft) [inline Tasks/UseDotNetV2/versioninstaller.ts:45]: this line will report the wrong url for 'could not download', which should be builds.dotnet.microsoft.com

**Comment by JamieMagee** [comment]
Related: https://github.com/dotnet/core/issues/9671

**Review by merlynomsft — APPROVED** [review]

**Review by tarunramsinghani — APPROVED** [review]

File: Tasks/UseDotNetV2/versioninstaller.ts (hunk 9 above)
Comment (sanjuyadav24) [reply Tasks/UseDotNetV2/versioninstaller.ts]: updated

**Pull request merged**